### PR TITLE
added skip_upload option 'upload' command

### PR DESCRIPTION
**Pull request by claasd**
This way, if there is an error in one of the packages, it can be checked
before uploading anything.
Additionally, for large packages the comression step can be done before
uploading.

### CODE REVIEW

File: conans/client/remote_manager.py [hunk 1]
```diff
@@ -90,15 +90,19 @@ def upload_package(self, package_reference, remote, retry, retry_wait):
             self._output.rewrite_line("Package integrity OK!")
         self._output.writeln("")
         logger.debug("====> Time remote_manager check package integrity : %f" % (time.time() - t1))
-
         the_files = compress_package_files(files, package_folder, self._output)
+        if not dry_run:
+
+            tmp = self._call_remote(remote, "upload_package", package_reference, the_files,
+                                    retry, retry_wait)
+            print(("WHAT2!"))
```
Comment (memsharded) [inline conans/client/remote_manager.py]: Please remove this print

File: conans/client/remote_manager.py [hunk 2]
```diff
@@ -35,24 +35,24 @@ def upload_conan(self, conan_reference, remote, retry, retry_wait, ignore_delete
...
+            return ret
         else:
-            msg += ": %s" % remote.url
-        self._output.info(msg)
+            return None
 
-        return ret
-
-    def upload_package(self, package_reference, remote, retry, retry_wait):
+    def upload_package(self, package_reference, remote, retry, retry_wait, dry_run = None):
```
Comment (memsharded) [inline conans/client/remote_manager.py]: For default params, the style is without spaces, like ``dry_run=None``. Also, if it is a boolean, you might want to use ``False`` as default

File: conans/client/manager.py [hunk 3]
```diff
@@ -492,9 +492,8 @@ def build(self, conanfile_path, current_path, test=False, filename=None):
             raise ConanException("Unable to build it successfully\n%s" % '\n'.join(trace[3:]))
 
     def upload(self, conan_reference_or_pattern, package_id=None, remote=None, all_packages=None,
-               force=False, confirm=False, retry=0, retry_wait=0):
+               force=False, confirm=False, retry=0, retry_wait=0, dry_run = None):
```
Comment (memsharded) [inline conans/client/manager.py]: Same as below, please use ``dry_run=False`` without spaces

**Comment by lasote** [comment]
> if there is an error in one of the packages, it can be checked before uploading anything.

What kind of error? 

> Additionally, for large packages the comression step can be done before
uploading.

Right, but if the only purpose is pre-compressing files maybe the "dry run" name is not the best, because actually is doing something, not only simulating. Maybe `--only-compress` or `skip-upload` or something similar? I don't know.

**Comment by claasd** [comment]
I think `--skip-upload` might be a good name. 
As to your other question, I observed some problems with package integrity. See the example below. The receipt and one package have been uploaded, then an error occurs:

```
D:\slave\ws\cab\extern\boost\1.63.0@0\ws>conan.exe upload "Boost/1.63.0@cab/extern" --all --confirm -r bop --retry 3 
Uploading Boost/1.63.0@cab/extern

Compressing recipe...                                                 
Uploading conanmanifest.txt                                           


Uploading conan_export.tgz                                            

Uploaded conan recipe 'Boost/1.63.0@cab/extern' to 'bop': http://conan.bop
Uploading package 1/5: 4f0fd3886115d238859ac2a3d41664e4236efc61

Checking package integrity...                                         
Package integrity OK!                                                 

Compressing package...                                                
Requesting upload permissions...                                      
Requesting upload permissions...Done!                                 

Uploading conan_package.tgz                                           

Uploading package 2/5: 5d34366328ff5fe8b9f93b912e5855ffb5fda596

Checking package integrity...                                         
Traceback (most recent call last):
  File "C:\Python34\Scripts\conan-script.py", line 9, in <module>
    load_entry_point('conan==0.19.2', 'console_scripts', 'conan')()
  File "C:\Python34\lib\site-packages\conans\conan.py", line 6, in run
    main(sys.argv[1:])
  File "C:\Python34\lib\site-packages\conans\client\command.py", line 1065, in main
    error = command.run(args)
  File "C:\Python34\lib\site-packages\conans\client\command.py", line 979, in run
    raise exc
  File "C:\Python34\lib\site-packages\conans\client\command.py", line 959, in run
    method(args[0][1:])
  File "C:\Python34\lib\site-packages\conans\client\command.py", line 826, in upload
    retry_wait=args.retry_wait)
  File "C:\Python34\lib\site-packages\conans\client\manager.py", line 540, in upload
    retry=retry, retry_wait=retry_wait)
  File "C:\Python34\lib\site-packages\conans\client\uploader.py", line 41, in upload_conan
    self._upload_conan(conan_ref, force, all_packages, retry, retry_wait)
  File "C:\Python34\lib\site-packages\conans\client\uploader.py", line 56, in _upload_conan
    retry, retry_wait)
  File "C:\Python34\lib\site-packages\conans\client\uploader.py", line 76, in upload_package
    self._remote_proxy.upload_package(package_ref, retry, retry_wait)
  File "C:\Python34\lib\site-packages\conans\client\proxy.py", line 295, in upload_package
    result = self._remote_manager.upload_package(package_ref, remote, retry, retry_wait)
  File "C:\Python34\lib\site-packages\conans\client\remote_manager.py", line 77, in upload_package
    if read_manifest.file_sums[fname] != expected_manifest.file_sums[fname]:
KeyError: 'include/boost/asio/ip/unicast.hpp'
```

**Comment by lasote** [comment]
Ok, I like `--skip-upload`, WDYT @memsharded?
About that error, it's interesting. I'm opening an issue to avoid the ugly crash but did you anything strange with that package like modify some file by hand or something?

**Comment by memsharded** [comment]
Yes, ``--skip-upload`` sounds good.

Please @claasd , this error is weird, can you provide more info about previous steps before that failure? Are you doing anything that could alter the package folder, so the manifest doesn't match? Basically the error is like the file ``include/boost/asio/ip/unicast.hpp`` from package binary has been somewhat removed from the package folder after being packaged (the moment where the manifest is computed)

**Comment by cawilson** [comment]
Is it possible to get this switch added to the docs? We were looking for this feature and only knew about it through luck in finding this PR 😃

**Comment by memsharded** [comment]
Added in https://github.com/conan-io/docs/commit/2b7d87d39e35531ce6059feb700531adef204166

Sometimes, it might be useful to run the ``conan upload --help`` to check current status. Sometimes it is difficult to keep the docs up to date, but the app --help is the truth.

**Pull request merged**